Fix writing of mixed content in JPZ clues.

jpd236 opened the pull request:
This offers a bit more flexibility to support JPZ files with clues that
are only partially formatted, e.g. "<i>italicized part</i> unformatted
part". While the spec suggests that mixed content is "not supported",
and indeed Puzzle Solver can't handle such content, the online solver at
Crossword Nexus does handle it, and without such support, we'd lose the
ability to have italicized titles in clues when importing .rgz files.

See #86

jpd236 commented:
The other option here would be to replace asterisks with quotes instead of italicizing them, but it seems like this is a good idea either way given that mixed content is supported when _reading_ JPZ files, so it should work symmetrically when writing them, even if Puzzle Solver doesn't support them.

The relevant code from the Crossword Nexus solver is here: https://github.com/crosswordnexus/html5-crossword-solver/blob/master/js/crosswords.js#L240

mrichards42 commented:
I'm not sure if that was an intentional decision originally or not, but I certainly don't mind breaking compatibility if there isn't a way around it.

Although according to the [spec](http://crossword.info/xml/rectangular-puzzle.xsd):

> In general can contain any html snippet, but at this time only 'span' 'b' 'i' 'sub' 'sup' are supported. Also the mixed content is not supported at this moment. So, all not-formated text should be wrapped into 'spans'. If all text has no format it can be stored without spans. 

so maybe we could get around it by doing this?
```xml
<i>italicized part</i><span> unformatted</span>
<!-- or would it be this? -->
<span><i>italicized part</i> unformatted</span>
```

jpd236 commented:
Ah - I had read that spec as suggesting the latter (which does not work), but I tried the former and it does appear to work fine with Puzzle Solver!

I'll work on updating the import script to surround the unformatted portions of clues that contain italics with span tags in the RGZ support PR. That said, I think this PR is still valid and even required, since with the code as currently written, it would only write the first chunk of the clue (e.g. whatever's in the first i/span tag).

mrichards42 commented:
So it looks like this is necessary to save formatted text that crossword solver will read. For crossword solver to read formatted clue text, we wold also need to do something like this, right?
```
for each node:
  if text node:
    append("<span>" + node + "</span>");
  else:
    append(node)
```

Clearly the shared xml ns is not the right place to do that, so maybe we want a jpz-specific version of SetInnerXML that would be used around here:
https://github.com/mrichards42/xword/blob/2b52c2fba841d91bfffafd0f6b1ce04eabeba35d/puz/formats/jpz/save_jpz.cpp#L239

(not that this needs to be done to merge this PR, I'm just trying to get the full picture)

jpd236 commented:
I'm not sure I follow your question. It seems like what you're asking is what to do when trying to save a clue like:

```plain text <tag>inner text</tag>```

because Crossword Solver wouldn't be able to read that. But I guess I'd argue that the input file was "wrong" in that case, because Crossword Solver wouldn't have been able to read it. And the logic in this PR should ensure that we write exactly what we read, rather than transform it in some way.

We could introduce <span> tags around plain text portions when saving if we wanted to be able to turn files which were previously not compatible with Crossword Solver into compatible files, I guess? But it seems a bit funny to me to mess with the input data like that.

Am I understanding right?

mrichards42 commented:
> But I guess I'd argue that the input file was "wrong" in that case, because Crossword Solver wouldn't have been able to read it. And the logic in this PR should ensure that we write exactly what we read, rather than transform it in some way.

I'm thinking more about converting files in other formats that support text styles (most likely ipuz, since we don't actually support saving ipuz right now) into jpz. `Puzzle` supports more or less a superset of the features of the file types we know how to load. Ideally we could read an ipuz with a mixed content clue and save it as a (valid) jpz.

jpd236 commented:
Added another commit to handle the ipuz -> jpz case. To keep the jpz-specific logic out of the shared XML functions, I added an overload of SetInnerXML which takes a pointer to a function that appends a child to a parent (and is also provided the child count for reference). The JPZ format wraps text children in spans as long as there's more than one child.

Checking child count isn't strictly necessary, but it seemed preferable to me since it's quite common that a clue is just plain text, and wrapping all such clues in a span seems unnecessary.

mrichards42 commented:
Looks good, thanks!
